epochStart/metachain/pendingMiniBlocks.go

```diff
@@ -1,257 +1,126 @@
 				continue
 			}
 
-			crossShard[string(mbHeader.Hash)] = mbHeader
-		}
-	}
-
-	return crossShard
-}
-
-func (p *pendingMiniBlockHeaders) getLastUsedMetaBlockFromShardHeaders(
-	lastNotarizedHeaders []data.HeaderHandler,
-) (map[string]*block.MetaBlock, error) {
-	mapLastUsedMetaBlocks := make(map[string]*block.MetaBlock)
-	for _, header := range lastNotarizedHeaders {
-		shardHdr, ok := header.(*block.Header)
-		if !ok {
-			return nil, epochStart.ErrWrongTypeAssertion
+			crossShardMiniBlocks[string(mbHeader.Hash)] = mbHeader
 		}
-
-		numMetas := len(shardHdr.MetaBlockHashes)
-		if numMetas == 0 {
-			continue
-		}
-
-		lastMetaBlockHash := shardHdr.MetaBlockHashes[numMetas-1]
-		if _, hashInMap := mapLastUsedMetaBlocks[string(lastMetaBlockHash)]; hashInMap {
-			continue
-		}
-
-		lastMetaHdr, err := p.getMetaBlockByHash(lastMetaBlockHash)
-		if err != nil {
-			return nil, err
-		}
-
-		mapLastUsedMetaBlocks[string(lastMetaBlockHash)] = lastMetaHdr
-	}
-
-	return mapLastUsedMetaBlocks, nil
-}
-
-func (p *pendingMiniBlockHeaders) getMetaBlockByHash(metaHash []byte) (*block.MetaBlock, error) {
-	peekedData, _ := p.metaBlockPool.GetHeaderByHash(metaHash)
-	metaHdr, ok := peekedData.(*block.MetaBlock)
-	if ok {
-		return metaHdr, nil
-	}
-
-	buff, err := p.metaBlockStorage.Get(metaHash)
-	if err != nil {
-		return nil, err
-	}
-
-	var metaHeader block.MetaBlock
-	err = p.marshalizer.Unmarshal(&metaHeader, buff)
-	if err != nil {
-		return nil, err
 	}
 
-	return &metaHeader, nil
+	return crossShardMiniBlocks
 }
 
 // AddProcessedHeader will add all miniblocks headers in a map
-func (p *pendingMiniBlockHeaders) AddProcessedHeader(handler data.HeaderHandler) error {
-	if check.IfNil(handler) {
+func (p *pendingMiniBlocksHeaders) AddProcessedHeader(headerHandler data.HeaderHandler) error {
+	if check.IfNil(headerHandler) {
 		return epochStart.ErrNilHeaderHandler
 	}
 
-	metaHdr, ok := handler.(*block.MetaBlock)
+	metaBlock, ok := headerHandler.(*block.MetaBlock)
 	if !ok {
 		return epochStart.ErrWrongTypeAssertion
 	}
 
-	crossShard := p.getAllCrossShardMiniBlocks(metaHdr)
+	crossShardMiniBlocks := p.getAllCrossShardMiniBlocks(metaBlock)
 
-	var err error
 	p.mutPending.Lock()
-	defer func() {
-		p.mutPending.Unlock()
-		if err != nil {
-			_ = p.RevertHeader(handler)
-		}
-	}()
-
-	for key, mbHeader := range crossShard {
-		if _, ok = p.mapMiniBlockHeaders[key]; !ok {
-			p.mapMiniBlockHeaders[key] = mbHeader
+	for mbHash, mbHeader := range crossShardMiniBlocks {
+		if _, ok = p.mapMiniBlockHeaders[mbHash]; !ok {
+			p.mapMiniBlockHeaders[mbHash] = mbHeader
 			p.mapShardNumMiniBlocks[mbHeader.ReceiverShardID]++
 			continue
 		}
 
-		delete(p.mapMiniBlockHeaders, key)
-		p.decrementNumMiniBlocks(mbHeader.ReceiverShardID)
-
-		var buff []byte
-		buff, err = p.marshalizer.Marshal(&mbHeader)
-		if err != nil {
-			return err
+		delete(p.mapMiniBlockHeaders, mbHash)
+		if p.mapShardNumMiniBlocks[mbHeader.ReceiverShardID] > 0 {
```

Comment:
can we log an error in case p.mapShardNumMiniBlocks[] == 0 and we deleted?

so I would instead do: if p.mapShard.....[] == 0{
log error
continue
}
p.mapShardNum....[]--

although we should afterwards pay attention to this kind of errors in the logs...

Reply:
Done